File: claw-os-sdk/go/tools.go
// Tool helper for Claw OS Go apps.
//
// Apps that want to fulfil a model-proposed tool call (returned in
// AiResponse.ToolCalls after Chat(..., ChatOptions{Tools: ...})) shell
// out through this file to `cos ai tool <name> --app <id> --args
// <json>`. The kernel resolves the name against the catalog, derives
// the caps verb + scope, runs caps::require under the app's own grants,
// executes the implementation, and writes an audit row.

package clawossdk

import (
	"encoding/json"
	"os"
	"strings"
)

// ToolUnavailableError: the `cos` binary could not be invoked or
// returned garbage (transport failure).
type ToolUnavailableError struct{ Msg string }

func (e *ToolUnavailableError) Error() string { return e.Msg }

// ToolDeniedError: a gate (capability / unknown tool / args shape)
// refused the call. Payload holds the structured kernel envelope.
type ToolDeniedError struct{ Payload map[string]any }

func (e *ToolDeniedError) Error() string {
	if m, ok := e.Payload["error"].(string); ok && m != "" {
		return m
	}
	return "Tool call denied"
}

// ToolResult is the kernel-mediated result of one tool invocation.
// Value is the JSON the catalog implementation produced (per-tool shape).
type ToolResult struct {
	Name   string
	AppID  string
	Status string
	Value  any
	Raw    map[string]any
}

// CatalogEntry is one row from `cos ai tools`.
type CatalogEntry struct {
	Name          string
	Summary       string
	Verb          string
	Stability     string
	ArgsSchema    map[string]any
	ReturnsSchema map[string]any
	Raw           map[string]any
}

// CallTool invokes a catalog tool through the kernel. args may contain any
// JSON value; nil is encoded as explicit JSON null. appID defaults to
// $COS_APP_ID. Returns a *ToolDeniedError for anything the
// gate refused, or *ToolUnavailableError for transport problems.
func CallTool(name string, args any, appID string) (*ToolResult, error) {
	if strings.TrimSpace(name) == "" {
		return nil, &ToolUnavailableError{Msg: "CallTool: name must be non-empty"}
	}
	app := appID
	if app == "" {
		app = os.Getenv("COS_APP_ID")
	}
	if app == "" {
		return nil, &ToolUnavailableError{Msg: name + ": app_id is required (pass appID or set COS_APP_ID)"}
	}
	payload, err := json.Marshal(args)
	if err != nil {
		return nil, &ToolUnavailableError{Msg: "Tool: could not encode args: " + err.Error()}
	}

	out, err := cosCallJSON("cos ai tool "+name, []string{"ai", "tool", name, "--app", app, "--args", string(payload)})
	if err != nil {
		if denied, ok := err.(*DeniedError); ok {
			return nil, &ToolDeniedError{Payload: denied.Payload}
		}
		if ue, ok := err.(*UnavailableError); ok {
			return nil, &ToolUnavailableError{Msg: ue.Msg}
		}
		return nil, err
	}
	if err := ValidateTool(out.Data); err != nil {
		return nil, &ToolUnavailableError{Msg: "tool result decode failed: " + err.Error()}
	}
	env := asMap(out.Data)
	return &ToolResult{
		Name:   asString(env["tool"]),
		AppID:  asString(env["app_id"]),
		Status: asString(env["status"]),
		Value:  env["result"],
		Raw:    env,
	}, nil
}

// Catalog returns the live tool catalog as exposed by `cos ai tools`.
// Apps shouldn't hard-code tool names without consulting this list; the
// catalog evolves and a tool can be deprecated or renamed.
func Catalog() ([]CatalogEntry, error) {
	out, err := cosCallJSON("cos ai tools", []string{"ai", "tools"})
	if err != nil {
		if denied, ok := err.(*DeniedError); ok {
			return nil, &ToolDeniedError{Payload: denied.Payload}
		}
		if ue, ok := err.(*UnavailableError); ok {
			return nil, &ToolUnavailableError{Msg: ue.Msg}
		}
		return nil, err
	}
	if err := ValidateToolCatalog(out.Data); err != nil {
		return nil, &ToolUnavailableError{Msg: "catalog decode failed: " + err.Error()}
	}
	env := asMap(out.Data)
	rawRows := env["tools"].([]any)
	entries := make([]CatalogEntry, 0, len(rawRows))
	for _, r := range rawRows {
		row := r.(map[string]any)
		entries = append(entries, CatalogEntry{
			Name:          asString(row["name"]),
			Summary:       asString(row["summary"]),
			Verb:          asString(row["verb"]),
			Stability:     asString(row["stability"]),
			ArgsSchema:    asMap(row["args_schema"]),
			ReturnsSchema: asMap(row["returns_schema"]),
			Raw:           row,
		})
	}
	return entries, nil
}

// ForChat normalises tool names for ChatOptions.Tools: trims whitespace
// and drops empties, so ForChat("fs.read_text", " kv.get ", "") yields
// two clean entries.
func ForChat(names ...string) []string {
	out := make([]string, 0, len(names))
	for _, n := range names {
		s := strings.TrimSpace(n)
		if s != "" {
			out = append(out, s)
		}
	}
	return out
}

File: claw-os-sdk/node/src/tools.ts
// Tool helper for Claw OS Node apps.
//
// Apps that want to fulfil a model-proposed tool call (returned in
// `ai.AiResponse.toolCalls` after `ai.chat(..., { tools: [...] })`)
// shell out through this module to `cos ai tool <name> --app <id>
// --args <json>`. The kernel resolves the name against the catalog,
// derives the caps verb + scope, runs caps::require under the app's own
// grants, executes the implementation, and writes an audit row.
//
//   import { ai, tools } from "@claw-os/sdk";
//   const proposal = ai.chat("Summarise /etc/hostname", {
//     tools: tools.forChat("fs.read_text"),
//   });
//   for (const c of proposal.toolCalls) {
//     const r = tools.call(c.name, c.input);
//   }

import { BridgeError, Unavailable, asObject, cosCallJson, hasError } from "./transport";
import {
  WireDecodeError,
  validateTool,
  validateToolCatalog,
} from "./generated";

/** Base class for every error this module raises. */
export class ToolError extends BridgeError {}
/** The `cos` binary could not be invoked or returned garbage. */
export class ToolUnavailable extends ToolError {}

/** A gate (capability / unknown tool / args shape) refused the call.
 * `payload` is the structured envelope the kernel returned. */
export class ToolDenied extends ToolError {
  readonly payload: Record<string, unknown>;
  constructor(payload: Record<string, unknown>) {
    super(String(payload["error"] ?? "Tool call denied"));
    this.payload = payload;
  }
}

/** The kernel-mediated result of one tool invocation. `value` is the
 * JSON the catalog implementation produced (shape is per-tool). */
export interface ToolResult {
  name: string;
  appId: string;
  status: string;
  value: unknown;
  raw: Record<string, unknown>;
}

/** One row from `cos ai tools`. */
export interface CatalogEntry {
  name: string;
  summary: string;
  verb: string;
  stability: string;
  argsSchema?: Record<string, unknown>;
  returnsSchema?: Record<string, unknown>;
  raw: Record<string, unknown>;
}

/**
 * Invoke a catalog tool through the kernel.
 *
 * Throws {@link ToolDenied} for anything the gate refused (unknown
 * tool, missing capability, malformed args) and {@link ToolUnavailable}
 * for transport problems.
 */
export function call(
  name: string,
  args: Record<string, unknown> = {},
  opts: { appId?: string } = {},
): ToolResult {
  if (!name || typeof name !== "string") {
    throw new ToolError("call: name must be a non-empty string");
  }
  const app = opts.appId || process.env.COS_APP_ID;
  if (!app) {
    throw new ToolError(`${name}: app_id is required (pass appId or set COS_APP_ID)`);
  }

  const argv = ["ai", "tool", name, "--app", app, "--args", JSON.stringify(args ?? {})];
  let outcome;
  try {
    outcome = cosCallJson(`cos ai tool ${name}`, argv);
  } catch (e) {
    if (e instanceof Unavailable) throw new ToolUnavailable(e.message);
    throw e;
  }
  if (outcome.status !== 0 || hasError(outcome.envelope)) {
    throw new ToolDenied(asObject(outcome.envelope));
  }
  try {
    validateTool(outcome.envelope);
  } catch (error) {
    if (error instanceof WireDecodeError) {
      throw new ToolUnavailable(`tool result decode failed: ${error.message}`);
    }
    throw error;
  }
  const env = outcome.envelope;
  return {
    name: env.tool,
    appId: env.app_id,
    status: env.status,
    value: env.result,
    raw: env,
  };
}

/**
 * Return the live tool catalog as exposed by `cos ai tools`.
 *
 * Apps shouldn't hard-code tool names without consulting this list; the
 * catalog evolves and a tool can be deprecated or renamed between
 * releases.
 */
export function catalog(): CatalogEntry[] {
  let outcome;
  try {
    outcome = cosCallJson("cos ai tools", ["ai", "tools"]);
  } catch (e) {
    if (e instanceof Unavailable) throw new ToolUnavailable(e.message);
    throw e;
  }
  if (outcome.status !== 0 || hasError(outcome.envelope)) {
    throw new ToolDenied(asObject(outcome.envelope));
  }
  try {
    validateToolCatalog(outcome.envelope);
  } catch (error) {
    if (error instanceof WireDecodeError) {
      throw new ToolUnavailable(`catalog decode failed: ${error.message}`);
    }
    throw error;
  }
  const env = outcome.envelope;
  return env.tools.map((row) => ({
      name: row.name,
      summary: row.summary,
      verb: row.verb,
      stability: row.stability,
      argsSchema: row.args_schema,
      returnsSchema: row.returns_schema,
      raw: { ...row },
    }));
}

/**
 * Normalise tool names for `ai.chat`'s `tools` option: trim whitespace
 * and drop empties, so `forChat("fs.read_text", " kv.get ", "")`
 * collapses to two clean entries.
 */
export function forChat(...names: string[]): string[] {
  const out: string[] = [];
  for (const n of names) {
    if (typeof n !== "string") {
      throw new ToolError(`forChat: tool names must be strings, got ${String(n)}`);
    }
    const s = n.trim();
    if (s) out.push(s);
  }
  return out;
}
